engine/engine.go

```diff
@@ -169,18 +173,19 @@ func GetEngine() Engine {
 }
 
 // Resolve launches the asynchronous execution of a resolution, given its ID
-func (e Engine) Resolve(publicID string) error {
-	_, err := e.launchResolution(publicID, true)
+func (e Engine) Resolve(publicID string, sm *semaphore.Weighted) error {
+	_, err := e.launchResolution(publicID, true, sm)
 	return err
 }
 
 // SyncResolve launches the synchronous execution of a resolution, given its ID
-func (e Engine) SyncResolve(publicID string) (*resolution.Resolution, error) {
-	return e.launchResolution(publicID, false)
+func (e Engine) SyncResolve(publicID string, sm *semaphore.Weighted) (*resolution.Resolution, error) {
+	return e.launchResolution(publicID, false, sm)
 }
 
-func (e Engine) launchResolution(publicID string, async bool) (*resolution.Resolution, error) {
-
+func (e Engine) launchResolution(publicID string, async bool, sm *semaphore.Weighted) (*resolution.Resolution, error) {
+	e.wg.Add(1)
```

Comment:
This .Add() looks suspicious.

.Add() should be done before the statement creating the goroutine (like below).
So either there is already a wg.Add before the goroutine calling launchResolution and this wg.Add is superfluous or the wg.Add should be moved before the go statement

Reply:
It's a bit non-intuitive, but from what I see this wg.Add() protects the initialization steps within launchResolution itself. The async goroutine it spawns is represented by another wg.Add().

I'll merge as-is to release 1.6.0, but it would warrant a comment indeed.